### ISSUE TITLE
[css-animations-1] Font-relative units in keyframes

### ISSUE THREAD

**Issue by andruud** [w3c/csswg-drafts#3751]
What does `em` resolve against in the following animation?

```
@keyframes anim1 {
    from { font-size: 10px; width: 10em; }
    to   { font-size: 20px; width: 20em; }
}
```

In Chrome, `anim1` produces an animation for `width` in `em`-space, where `em` is resolved against the animated `font-size` for each frame.

In FF, `anim1` produces an animation for `width` from `160px` to `320px`, i.e. the `font-size` of the keyframe is ignored.

**Comment by emilio** [comment]
cc @birtles 

I _think_ off-hand I'd consider this a Firefox bug. But note that Blink has issues in here as well (i.e., I think the third element in this test-case should grow):

```html
<!doctype html>
<style>
@keyframes anim1 {
  from { width: 10em; font-size: 10px; }
  to   { width: 30em; font-size: 50px; }
}

@keyframes anim2 {
  from { font-size: 10px; width: 10em; }
  to   { font-size: 50px; width: 30em; }
}

@keyframes anim3 {
  from { font-size: 10px; }
  to   { font-size: 50px; }
}

div {
  height: 60px;
  background: purple;
  animation: anim1 10s linear infinite;
  margin: 5px 0;
}

div + div {
  animation-name: anim2;
}
div + div + div {
  width: 30em;
  animation-name: anim3;
}
</style>
<div></div>
<div></div>
<div></div>
```

**Comment by andruud** [comment]
Yeah, that's a known related issue in Blink.

**Comment by andruud** [comment]
Hmm. :thinking: 

§ 3. Animations [css-animations-1]:

> CSS Animations affect computed property values. This effect happens by adding a specified value to the CSS cascade ([CSS3CASCADE]) (at the level for CSS Animations) that will produce the correct computed value for the current state of the animation.

But then also:

§ 3. Combining Values: Interpolation, Addition, and Accumulation [css-values-4]

> These operations are only defined on computed values. (As a result, it is not necessary to define, for example, how to add a <length> value of 15pt with 5em since such values will be resolved to their canonical unit before being passed to any of the above procedures.)

So ... should the interpolation happen in `em` space, or not?

**Comment by SelenIT** [comment]
The newest definition of the `width` property that I could find (in [[css-sizing-3]](https://drafts.csswg.org/css-sizing-3/#propdef-width)) says that its animation type is "by computed value type", and computed value is "as specified, _with `<length-percentage>` values computed_". The [css-animations] spec seems to have no exact rules of length interpolation, instead it has an [issue](https://drafts.csswg.org/css-animations/#issue-73aacf21) linking to the [[css-transitions] rules](https://drafts.csswg.org/css-transitions/#application) as an example. These rules imply that interpolation happens between the start and end values only, i.e. Firefox's behavior seems to be correct according to them (though it might be counter-intuitive).

Interestingly, the MDN page for the `width` property currently [reports](https://developer.mozilla.org/en-US/docs/Web/CSS/width#Specifications) its animation type as "a length, percentage _or calc();_", apparently implying that `calc()` is a separate animation type, so interpolations between `10em` and `20em` and between `calc(10em)` and `calc(20em)` would probably happen in different spaces. But I haven't found any justification for such distinction in neither [[css-values-4]](https://drafts.csswg.org/css-values-4/#calc-calculation) nor animations nor transitions spec.

With all that said, I wouldn't consider either behavior a bug until the [css-animations] spec resolves its issue about the interpolation rules between keyframes.

**Comment by birtles** [comment]
So this is basically a dupe of #66 (which, if you follow far enough, will take you to [this mail from dbaron](https://lists.w3.org/Archives/Public/www-style/2014Aug/0132.html))

I wrote a test case with a few variations on the test case from the first comment: https://codepen.io/birtles/pen/drQQaJ.

Blink seems to mostly do G-β but has trouble with overlapping keyframes and, as Emilio points out,  reflecting the animated font-size on static declarations.

Edge doesn't seem to recognize `animation-timing-function` on keyframes when there are overlapping keyframes. I didn't test Emilio's example on Edge.

I _thought_ we fixed some of these issues in Firefox when we switched to Servo, but I guess not. I believe we might have deferred the ordering dependency introduced by `font-size` until we implement animation of custom properties since they introduce further ordering dependencies between properties.

So I think we already have a resolution to do G-β, it just needs to be implemented in Firefox, tested (the WPT for this spec need rewriting first) and specced. Unfortunately this behavior does mean that some animations can no longer effectively be run on the compositor.
